File: dag_sort.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <sstream>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Part {
public:
	string name() {
		return "X";
	}   
};

class Operation {
public:
	string name;

	Operation( ) : mDFSFlag(0) {}
	Operation( const Operation& operation) : name(operation.name),  mDFSFlag(0) { }
	Operation(string _name) : name(_name),  mDFSFlag(0) {
    }

	void operate(Part p) {
		cout << "Operation " << name << " on part " << p.name() << endl;
    }
    
    Operation& operator=(Operation other) {
        name = other.name;
        return *this;
    }
    void addEdge(Operation& operation){
        if(operation.name.compare(name) == 0) return;
        auto p = find (mParentSteps.begin(), mParentSteps.end(), operation.name);
        if(p == mParentSteps.end()) {
            mParentSteps.push_back(operation.name);
        }
	}

    void set_flag(int flag){
        mDFSFlag = flag;
    }

    int get_flag(){
        return mDFSFlag;
    }

    const vector<string>& get_next_steps() const{
        return mParentSteps;
    }
private:
    vector<string> mParentSteps; //for all dependencies
    int mDFSFlag;//0--not visited, 1--visited;
};

class StepManager {
public:
	void add_operation(Operation operation, vector<string> prerequisites) {
		// Add your implementation here. This method is called
		// once for each type of operation during setup.

        auto got = mOperations.find (operation.name);
        if(got == mOperations.end())//doesn't exists
            mOperations[operation.name] = operation;

        for(auto i:prerequisites){//construct the graph
           auto got = mOperations.find (i);
           if(got == mOperations.end())//doesn't exists, add it in.
                mOperations[i] = Operation(i);
            mOperations[operation.name].addEdge(mOperations[i]);
        }
	}

	void perform_operation(string operation_name, Part p) {
		// Add your implementation here. When this method is called,
		// you must call the 'operate()' method for the named operation,
		// as well as all prerequisites for that operation.

        if(mOperations.empty())
            return;

        //perform an dfs
        mOperationSequence.clear();
 
        //reset dfs flags to 0;
        for ( auto it = mOperations.begin(); it != mOperations.end(); ++it ){
            it->second.set_flag(0);
        }

        dependency_sort(operation_name);

        cout << "All Operations for final" << endl;
        //do operations in sequence, which are in reversed order
        for(auto o :mOperationSequence)
        {
            mOperations[o].operate(p);
        }
}
private:
    void dependency_sort(const string& name){
        if(mOperations[name].get_flag() != 0)//processed already
            return;
        cout <<"discover operation: " << name << endl;
        mOperations[name].set_flag(1);
        for( auto next : mOperations[name].get_next_steps()){
            Operation& operation = mOperations[next];
            if(operation.get_flag() == 0) {
                dependency_sort(operation.name);
            }
        }
        //update operationSequence;
        cout << "add operation: " << name << endl;
        mOperationSequence.push_back(name);
    }
    unordered_map<std::string,Operation> mOperations;
    vector<string> mOperationSequence;    
};

int main() {
	StepManager step_mgr;

	string s;
	while (getline(cin, s)) {
		if (s[0] == '#')
			continue;
        if(s.length() == 0)
            break;

		stringstream ss(s);
		string name, tmp;
		vector<string> prerequisites;

		getline(ss, name, ',');
		while (getline(ss, tmp, ',')) {
			prerequisites.push_back(tmp);
		}

		step_mgr.add_operation(Operation(name), prerequisites);
	}

	step_mgr.perform_operation("final", Part());
}

File: dag_sort.py
#dag sort in python
import sys
import fileinput

class Part:
    def name(self):
        return "X";
class Operation:

    def __init__(self, _name):
        self.dfsFlag = 0
        self.parent_steps = []
        self.name = _name;
    def addEdge(self, operation):
        if(operation.name == self.name): return

        if(not operation.name in self.parent_steps):
            self.parent_steps.append(operation.name)

    def operate(self, part):
        print "Operation %s on Part %s" % ( self.name, part.name())

class StepManager:
    def __init__(self):
        self.Operations = {}
        self.OperationSequence = []

    def add_operation(self, operation_name, prerequisites):
        if(not self.Operations.has_key(operation_name)):
            self.Operations[operation_name] = Operation(operation_name);

        for i in prerequisites:
            i = i.rstrip()
            if(not self.Operations.has_key(i)):
                self.Operations[i] = Operation(i)

            self.Operations[operation_name].addEdge(self.Operations[i])

    def perform_operation(self, operation_name,  p):
        if(len(self.Operations) == 0): return
        for val in self.Operations.values():
            val.dfsFlag = 0

        self.OperationSequence = []
        self.dependency_sort(operation_name)

        for i in self.OperationSequence:
            self.Operations[i].operate(p)

    def dependency_sort(self, operation_name):
        if(self.Operations[operation_name].dfsFlag != 0): return

        self.Operations[operation_name].dfsFlag = 1
        for i in self.Operations[operation_name].parent_steps:
            self.dependency_sort(i)

        self.OperationSequence.append(operation_name)

if __name__ ==  "__main__":
    step_mgr = StepManager()
    for line in fileinput.input():
        if(len(line) == 0): continue
        if(line[0] == '#'):
            continue

        operations =line.split(',')
        if(operations):
            prerequisites = []
            if(len(operations) >1):
                prerequisites = operations[1:]
            step_mgr.add_operation(operations[0].rstrip(), prerequisites)
    print step_mgr.Operations.keys()
    step_mgr.perform_operation("final", Part())
    step_mgr.perform_operation("nine", Part())


